### PR TITLE
SIP2-132: Implement refresh token rotation

### PR DESCRIPTION
**Pull request by gurleenkaurbp**
## Purpose
Upgrade sip2 module to the new RTR-based approach.
 

## Approach
The approach taken here is by leveraging the RTR implementation which is present in okapi-common. 
In Sip2 the login handler will used the Okapi-common's Client interface to bring back the access token based on the provided credentials. Before, any command that is being executed in sip2 the access token will be fetched based on the cache (if present) or else from the Okapi-Common's Client getToken.

 

## Pre-Merge Checklist:
 Before merging this PR, please go through the following list and take appropriate actions.

 - Does this PR meet or exceed the expected quality standards?
   - [x] Code coverage on new code is 80% or greater
   - [x] Duplications on new code is 3% or less
   - [x] There are no major code smells or security issues
 - Does this introduce breaking changes?
   - [x] Were any API paths or methods changed, added or removed?
   - [ ] Were there any schema changes?
   - [ ] Did any of the interface versions change?
   - [ ] Were permissions changed, added, or removed?
   - [ ] Are there new interface dependencies?
   - [ ] There are no breaking changes in this PR.
   - [x] Check logging.

 If there are breaking changes, please **STOP** and consider the following:

 - What other modules will these changes impact?
 - Do JIRAs exist to update the impacted modules?
   - [ ] If not, please create them
   - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
   - [ ] Do they have all they appropriate links to blocked/related issues?
 - Are the JIRAs under active development?
   - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
 - Do PRs exist for these changes?
   - [ ] If so, have they been approved?

 Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

 While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Review by pavankumar181 — APPROVED** [review]

**Review by nirmalr-shah — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -51,20 +51,20 @@ On Windows, inline JSON configuration is in double quotes and the inner double q
...
+| `path`                         |string| Path name of the tenant configuration file for file type stores.                                                                                                                                |
+| `optional`                     |boolean| If a failure is caught while loading the tenant configuration from an optional store, the failure is logged, but the processing does not fail. Instead, the tenant configuration will be empty. |
+| `netServerOptions`             |JSON object| Configuration options for the server. These are Vertx options and are numerous. See: [NetServerOptions](https://vertx.io/docs/apidocs/io/vertx/core/net/NetServerOptions.html).                 |
+| `token_cache_capacity`         |int| Max token cache size. Default size is 100.                                                                                                                                                      |                                                                                                                                                           |
```
Comment (steveellis) [inline README.md:67]: I think the whitespace added here might make the markdown wonky when it renders. Or will it be ignored?

File: src/main/java/org/folio/edge/sip2/repositories/LoginRepository.java [hunk 2]
```diff
@@ -43,31 +39,32 @@ public Future<LoginResponse> login(Login login, SessionData sessionData) {
...
+      log.error("Login does not have a valid authentication token");
+      sessionData.setAuthenticationToken(null);
+      return Future.succeededFuture(LoginResponse.builder().ok(FALSE).build());
+    }
 
-    final JsonObject credentials = new JsonObject()
-        .put("username", user)
-        .put("password", password);
+    return authToken
+     .compose(token -> {
```
Comment (steveellis) [inline src/main/java/org/folio/edge/sip2/repositories/LoginRepository.java:57]: Think you may need another space (or 3 spaces) before the block with the `.compose`.

File: src/main/java/org/folio/edge/sip2/repositories/PasswordVerifier.java [hunk 3]
```diff
@@ -54,17 +55,21 @@ public Future<PatronPasswordVerificationRecords> verifyPatronPassword(
...
+                    .passwordVerified(TRUE)
+                    .build());
+                } else {
+                  return Future.succeededFuture(PatronPasswordVerificationRecords.builder()
+                    .extendedUser(extendedUser)
+                    .errorMessages(Collections.singletonList(sessionData.getLoginErrorMessage()))
+                    .passwordVerified(FALSE)
+                    .build());
+                }
+
```
Comment (steveellis) [inline src/main/java/org/folio/edge/sip2/repositories/PasswordVerifier.java:71]: Can probably remove space.

File: src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java [hunk 4]
```diff
@@ -149,6 +155,11 @@ public Future<PatronStatusResponse> performPatronStatusCommand(
         sessionData.isPatronPasswordVerificationRequired());
 
     return passwordVerifier.verifyPatronPassword(patronIdentifier, patronPassword, sessionData)
+      .onFailure(throwable -> {
+        if (throwable instanceof ClientException) {
+          sessionData.setErrorResponseMessage(invalidPatron(patronStatus, FALSE).result());
+        }
+      })
         .compose(verification -> {
```
Comment (steveellis) [inline src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java:163]: Something is funky with the indents here. Maybe the block above needs to come forward a couple of spaces to line up with eh one below.

File: src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java [hunk 5]
```diff
@@ -196,6 +207,16 @@ public Future<EndSessionResponse> performEndPatronSessionCommand(
...
+        if (throwable instanceof ClientException) {
+          sessionData.setErrorResponseMessage(EndSessionResponse.builder()
+              .endSession(FALSE)
+              .transactionDate(OffsetDateTime.now(clock))
+              .institutionId(endPatronSession.getInstitutionId())
+              .patronIdentifier(endPatronSession.getPatronIdentifier())
+              .build());
+        }
+      })
         .map(verification -> EndSessionResponse.builder()
```
Comment (steveellis) [inline src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java:220]: Should the `.onFailure` line up with the `.map` indent?

File: src/test/java/org/folio/edge/sip2/handlers/CheckoutHandlerTests.java [hunk 6]
```diff
@@ -86,6 +87,61 @@ public void canExecuteASampleCheckoutUsingHandler(
         })));
   }
 
+  @Test
+  void cantExecuteASampleCheckoutUsingHandler(
+      @Mock CirculationRepository mockCirculationRepository,
+      Vertx vertx,
+      VertxTestContext testContext) {
```
Comment (steveellis) [inline src/test/java/org/folio/edge/sip2/handlers/CheckoutHandlerTests.java:94]: Sorry now I'm obsessing about spaces. Does this block need to come forward 1 space?

File: src/test/java/org/folio/edge/sip2/handlers/CheckoutHandlerTests.java [hunk 7]
```diff
@@ -86,6 +87,61 @@ public void canExecuteASampleCheckoutUsingHandler(
...
+        .build());
+
+    handler.execute(checkout, sessionData).onComplete(
+        testContext.failing(sipMessage -> testContext.verify(() -> {
+
+          assertEquals("Incorrect Username", sipMessage.getMessage());
+
+          testContext.completeNow();
+        })));
+  }
```
Comment (steveellis) [inline src/test/java/org/folio/edge/sip2/handlers/CheckoutHandlerTests.java:143]: I'm not completely sure how this test differs from the one below. I mean I can see the additional properties, and I can see the method name is different. Would some comments in the body of the test make it more clear what it is demonstrating?

File: src/test/java/org/folio/edge/sip2/handlers/PatronStatusHandlerTests.java [hunk 8]
```diff
@@ -98,7 +99,67 @@ void canExecutePatronStatus(Vertx vertx,
         })));
   }
 
+  @Test
+  void cantExecutePatronStatus(Vertx vertx,
+                              VertxTestContext testContext,
+                              @Mock PatronRepository mockPatronRepository) {
```
Comment (steveellis) [inline src/test/java/org/folio/edge/sip2/handlers/PatronStatusHandlerTests.java:105]: This style of indents is different from the other tests. Would it be good to make them all the same? I like this style which intellij will help with, but this one needs to come forward 1 space.

File: src/test/java/org/folio/edge/sip2/handlers/RenewAllHandlerTests.java [hunk 9]
```diff
@@ -65,18 +66,18 @@ public void canRenewAllWithHandler(Vertx vertx,
         .build()
       ));
 
-    
-    
-    
+
+
+
```
Comment (steveellis) [inline src/test/java/org/folio/edge/sip2/handlers/RenewAllHandlerTests.java:71]: Maybe don't need so many new lines here.

File: src/test/java/org/folio/edge/sip2/handlers/RenewAllHandlerTests.java [hunk 10]
```diff
@@ -87,5 +88,56 @@ public void canRenewAllWithHandler(Vertx vertx,
 
 
   }
-  
+
+  @Test
+  void cantRenewAllWithHandler(Vertx vertx,
+                                     VertxTestContext testContext,
+                                     @Mock CirculationRepository mockCirculationRepository) {
```
Comment (steveellis) [inline src/test/java/org/folio/edge/sip2/handlers/RenewAllHandlerTests.java:95]: Indents probably need to be made consistent.

File: src/test/java/org/folio/edge/sip2/handlers/RenewHandlerTests.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        .ok(TRUE)
+        .renewalOk(TRUE)
+        .transactionDate(OffsetDateTime.now())
+        .institutionId("diku")
+        .build()
+      ));
+
+
+
+
```
Comment (steveellis) [inline src/test/java/org/folio/edge/sip2/handlers/RenewHandlerTests.java:69]: Don't need so many new lines here

File: src/test/java/org/folio/edge/sip2/handlers/RenewHandlerTests.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    handler.execute(renew, sessionData).onComplete(
+        testContext.succeeding(sipMessage -> testContext.verify(() -> {
+          assertNotNull(sipMessage);
+          assertEquals(expectedString, sipMessage);
+          testContext.completeNow();
+        }
+    )));
+
+
```
Comment (steveellis) [inline src/test/java/org/folio/edge/sip2/handlers/RenewHandlerTests.java:87]: Extra newline

**Review by steveellis — APPROVED** [review]
Lots of great refactoring here Gurleen! Most of my comments are just formatting.

File: src/main/java/org/folio/edge/sip2/repositories/LoginRepository.java (hunk 2 above)
Comment (gurleenkaurbp) [reply src/main/java/org/folio/edge/sip2/repositories/LoginRepository.java:57]: CheckStyle plugin does not agree to this :) the build fails if I increase/decrease the space count unfortunately.

File: src/main/java/org/folio/edge/sip2/repositories/PasswordVerifier.java (hunk 3 above)
Comment (gurleenkaurbp) [reply src/main/java/org/folio/edge/sip2/repositories/PasswordVerifier.java:71]: Checkstyle stuff, cant do much here.

File: src/test/java/org/folio/edge/sip2/handlers/CheckoutHandlerTests.java (hunk 7 above)
Comment (gurleenkaurbp) [reply src/test/java/org/folio/edge/sip2/handlers/CheckoutHandlerTests.java:143]: This test code covers the code of lines present in CheckOutHandler class when the ClientExcecption is thrown with the error message "Incorrect Username."
The incorrectUsernameFailure test case is for EndPatronSessionHandler.

File: src/test/java/org/folio/edge/sip2/handlers/RenewAllHandlerTests.java (hunk 9 above)
Comment (gurleenkaurbp) [reply src/test/java/org/folio/edge/sip2/handlers/RenewAllHandlerTests.java:71]: Done

File: src/test/java/org/folio/edge/sip2/handlers/RenewHandlerTests.java (hunk 11 above)
Comment (gurleenkaurbp) [reply src/test/java/org/folio/edge/sip2/handlers/RenewHandlerTests.java:69]: Done

**Comment by gurleenkaurbp** [comment]
> Lots of great refactoring here Gurleen! Most of my comments are just formatting.

Thanks steve, unfortunately the checkStyle plugin has made me induce those indentations and tab spaces without which the build fails.

File: README.md (hunk 1 above)
Comment (gurleenkaurbp) [reply README.md:67]: Modified and brought it back to its original state with just an extra Config option inclusion.

**Pull request merged**